Multi-window backtesting for TimeSeriesPredictor

shchur opened the pull request:
Currently, if no `tuning_data` is passed to `autogluon.timeseries.TimeSeriesPredictor.fit`, we use the last `prediction_length` timesteps of `train_data` as hold-out validation set. This PR enables new train / val splitting strategies for timeseries datasets, such as multi-window backtesting.

- New abstract class `autogluon.timeseries.splitter.AbstractTimeSeriesSplitter` that encapsulates the train / val split logic in its `split` method.
- Two currently supported strategies:
  - `LastWindowSplitter` - reserve last `prediction_length` steps (default, same as before)
  - `SlidingWindowSplitter` - reserve multiple windows of length `prediction_length` (potentially overlapping) for validation
- `autogluon.timeseries.TimeSeriesPredictor` now accepts an optional argument `splitter`.
- `tuning_data.index` is now different from `train_data.index` (a suffix `_[{start}:{end}]` is appended to show how the original series was split). We do this because a single item in `train_data` may now correspond to multiple items in `tuning_data`.  This shouldn't affect users since `tuning_data` is not exposed in the public API.

----
Updates:
- `autogluon.timeseries.dataset.ts_dataframe.TimeSeriesDataFrame` now makes sure that the provided `item_id` index either has all string or all integer types.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

shchur commented on timeseries/src/autogluon/timeseries/splitter.py:
```diff
@@ -0,0 +1,192 @@
+
+        For example, the series `val_data.loc["0_[None:-10]"]` in the validation dataset can equivalently be
+        obtained as `ts_dataframe.loc[0][None:-10]`.
+
```
One problem with this approach is that we cannot determine if the original `item_id` was an `int` or a `str` anymore after obtaining the new ID as `f"{idx}_[{start}:{end}]`. 

This may be slightly problematic if, for example, we at some point want to retrieve the respective local model trained on `train_data` and score it on a sequence from `tuning_data` with the new index.

The only alternatives I could think of are
1. Converting the `item_id` index of `train_data` to `str` as well
2. Adding an additional index level that denotes the split (in addition to the `item_id` and `timestamp` levels)

Both of these, however, look really dangerous and can have side effects.

shchur replied:
I guess this shouldn't be a problem anymore with the new implementation for local models that we discussed.

shchur commented on timeseries/src/autogluon/timeseries/splitter.py:
```diff
@@ -0,0 +1,192 @@
+        item_id, timestamp = multiindex_element
+        return (f"{item_id}{suffix}", timestamp)
+
+    result = ts_dataframe.copy(deep=False)
```
Something bad may happen if a model tries to modify the data in-place. Should we 
1. hope that this never happens, or
2. be extra safe, do a deep copy and waste memory?

shchur commented on timeseries/src/autogluon/timeseries/splitter.py:
```diff
@@ -0,0 +1,192 @@
+        train_data = ts_dataframe.slice_by_timestep(slice(None, -prediction_length))
+
+        return train_data, pd.concat(validation_dataframes)
+
```
At some point after benchmarking the current version I will implement an `AutomaticSplitter` that determines `num_windows` and `overlap` based on some heuristic.

shchur commented on timeseries/src/autogluon/timeseries/splitter.py:
```diff
@@ -0,0 +1,192 @@
+        if num_too_short_series > 0:
+            logger.warning(f"{num_too_short_series} are too short and won't appear in the training set")
+        if num_too_short_series == ts_dataframe.num_items:
+            raise ValueError(f"{self.name} produced an empty training set since all sequences are too short")
```
Does it make sense to completely remove the series that are too short to produce `num_windows` many prediction intervals (as current version of the code does), or maybe we should make sure that, e.g., at least 50% of original series's length is preserved for training?

canerturkmen replied:
I'm OK with this logic, but I can't see where the too short time series disappear. Does `slice_by_timestep` handle this gracefully?

canerturkmen commented on timeseries/tests/unittests/test_splitter.py:
```diff
@@ -0,0 +1,104 @@
+        assert expected_length == new_length
+
+    # Validation series are as long as implied by their item_id
+    for new_item_id in val_data.iter_items():
```
we can follow one assert per test here?

canerturkmen reviewed (commented):
looks great in general! just a few minor comments.

shchur commented on timeseries/src/autogluon/timeseries/dataset/ts_dataframe.py:
```diff
@@ -229,9 +231,15 @@ def _validate_multi_index_data_frame(cls, data: pd.DataFrame):
+        item_id_is_string = (item_id_index == item_id_index.astype(str)).all()
+        item_id_is_int = pd.api.types.is_integer_dtype(item_id_index)
+        if not (item_id_is_string or item_id_is_int):
+            raise ValueError(f"all entries in index `{ITEMID}` must be of integer or string dtype")
```
We now always ensure that the `ITEMID` columns of the index either contains all integers or all strings. This will avoid potential problems, such as collision of two time series with `item_id`s `0` (int) and `"0"` (str).

shchur commented:
Thanks @canerturkmen! I've addressed the comments and additionally added a check to `TimeSeriesDataFrame` that ensures that the `item_id` index contains only integer or only string values. This should prevent us from getting duplicate `item_id`s in the validation set.

canerturkmen reviewed (approved):
looks great! thanks!
